sphinx/ext/autodoc/__init__.py

```diff
@@ -899,6 +902,15 @@ def generate(self, more_content: Any = None, real_modname: str = None,
             except PycodeError:
                 pass
 
+        try:
+            subject = inspect.unpartial(self.object)
+            self.canonical_objpath = subject.__qualname__.split(".")
+            self.canonical_fullname = '%s::%s' % (self.real_modname, subject.__qualname__)
```

Comment:
Could you use `safe_getattr` instead of accessing `__qualname__` attribute directly? Users sometimes pass a strange object to autodoc. So we have to use it to access attributes.